### DISCUSSION TITLE
Entity Expansion in the XML parser

### DISCUSSION THREAD

**Discussion by Devasta** [ballsteve/xrust#26] [Planning]
Howya Steve,

Want to pick your brain for a bit if I can, I've hit a bit of a roadblock.

So, parsing DTDs is mostly grand, I've got something together that parses the various bits out (though it doesn't do much with element or attrlist declarations, that can be next months project!), however I am struggling to work out how to do the entity expansion.

I was hoping to do a simple approach of straight insertion of the entity onto the front of the input string, then letting the parser carry on as normal (lets forget the billion laugh attach for now), however I am wrestling with the ownership model in rust at the moment: it simply will not allow amendments to a reference, nor allow you return a reference to a string where the owner is the function itself.
```

fn myfun(input: &str) -> &str {
    let mystr = "FIRSTPART";
    return &(mystr.to_string() + input);
}

fn main() {
    let i = "SECONDPART";
    let q = myfun(&i);
    println!("{}",q)
}

```
> error[E0515]: cannot return reference to temporary value
>  --> src\main.rs:4:12
>   |
> 4 |     return &(mystr.to_string() + input);
>   |            ^---------------------------
>   |            ||
>   |            |temporary value created here
>   |            returns a reference to data owned by the current function

Above is a very simple example of the error from the rust compiler. The fix appears to be to return a String instead of a &str, however all subsequent parts of the parser expect a &str.

I was thinking the solution might be to do a sweep with a hand written parser over the DTD, pull in the entities and then a second sweep over everything once we had the list, dealing with it like we do the &gt; and &lt; characters, however since the entities can contain XML that needs parsing, that won't work either.

Any idea what approach I should take? This has me totally stumped.

**Comment by ballsteve** [comment]
Hi Daniel,

Indeed, this is a wicked problem. I guess that’s why JSON is supported by Serde, etc, rather than XML.

It would be good if nom could switch buffers, and then switch back when the entity body is finished. Perhaps you need to create a sub parser? Look at the do_parse macro, or the FromStr trait?

Perhaps we need to reach out to geal?

I’ve encountered a few situations where I need to mostly read data, but just occasionally modify it. To solve this I’ve used a lot of interior mutability (RefCell). However, that has a runtime performance penalty. The changes to the code I’m working on at the moment involve redesigning the code to eliminate that approach, without much success so far :-(

There are a couple of places where I use a 2-pass approach. The first pass figures out what needs to be done, and then the second pass does the work. You may need to do that in this case; the first pass just deals with DTDs and expanding parameter entities, the second pass parses the XML. However, dealing with parameter entity expansion may need to be iterative, multi-stage itself, since a parameter entity may reference other parameter entities (this must eventually terminate, since there are not allowed to be cycles). So this stage returns a tuple (bool, String), where the bool indicates whether an expansion occurred. The process would be related until the bool is false.

Ultimately, we could simply not support DTDs (or at least parameter entities) and use XInclude for composite documents. Do you need parameter entities, as opposed to DTD validation?

I’ll have a think about it and get back to you,
Steve Ball

**Comment by ballsteve** [comment]
Have you looked at nom_supreme? It has parse_from_str and FromStrParser functions that might help.

**Comment by Devasta** [comment]
I'll take a look into that, thanks!

To be honest I personally never use DTDs, but my job is always outbound data reporting, never really deal with inbound data. I figure its worth trying to support if I can, you mentioned previously that docbook uses them?

Whatever the solution is, it'll need to work for both parameter entities and internal entities, we have the predefined ones working fine because they are always character data, if XML gets included in one of those entities then it'll also need parsing.

**Comment by ballsteve** [comment]
It would be easier to solve this problem if nom supported passing additional arguments to the parser. This would allow a parser  context to be passed in which could have a HashMap to store and retrieve the entity declarations.

However, in the absence of that a multi-pass solution should be possible. I wrote an entity expander on Friday, see below, that seems to work without resorting to trickery like interior mutability.

I'll think about it further...

use std::collections::HashMap;
 
enum S {
    Literal(String),
    Ref(&'static str),
}
 
struct VS(Vec<S>);
 
impl VS {
    fn resolve(&self, hm: &HashMap<&str, VS>) -> String {
        self.0.iter().fold(String::from(""), |mut acc, s| match s {
            S::Literal(t) => {
                acc.push_str(t.as_str());
                acc
            }
            S::Ref(r) => {
                acc.push_str(hm.get(r).unwrap().resolve(hm).as_str());
                acc
            }
        })
    }
}
fn main() {
    let mut hm: HashMap<&str, VS> = HashMap::new();
    hm.insert("pfft", VS(vec![S::Literal(String::from("pfft"))]));
    hm.insert(
        "foo",
        VS(vec![
            S::Literal(String::from("foo")),
            S::Ref("pfft"),
            S::Literal(String::from("foo")),
        ]),
    );
    hm.insert(
        "bar",
        VS(vec![
            S::Literal(String::from("bar")),
            S::Ref("foo"),
            S::Literal(String::from("bar")),
        ]),
    );
    let s: VS = VS(vec![
        S::Literal(String::from("start")),
        S::Ref("bar"),
        S::Literal(String::from("end")),
    ]);
    println!("result: {}", s.resolve(&hm));
}

**Comment by ballsteve** [comment]
Hi Daniel,

Now that I've completed work on the Forest module, I thought I'd take a look at the entity expansion problem.

Have you made any progress? What's the status?

Cheers,
Steve

**Comment by Devasta** [comment]
Hi Steve,

I'm afraid I've not made any progress on this, I've been wrestling with the borrow checker and not had any breakthrough, and a recent promotion at work has been consuming a lot of my time the past few weeks (a project should be close to finishing up in June though, so definitely will have more time to look at it then).

https://github.com/Devasta/XRPC

I have been messing about trying out some different things; I was mulling over reimplementing the parser combinators we use but using a numeric offset rather than having the string passed along like nom does it, hoping that that would work around the ownership rueles, but i've not yet confirmed whether or not its possible to do the entity expansion this way either (and even if it is, I am fairly certain it wouldn't be at all efficient for large documents, it's something we'd probably just fall back on if necessary).

We might need to do a multi-sweep approach, (maybe even have the second sweep be an XSLT that runs through the document filling in the expanded entities!), and then possibily revisiting if we decide we want to implement streaming?

**Comment by ballsteve** [comment]
Given the requirement to do entity expansion, I doubt a non-destructive parser would be feasible. But I could be wrong ;-)

I think a multi-stage approach is needed. The current implementation has multiple stages anyway. The main thing is to get the functionality correct and complete; performance optimisation can come after that.

I'll take a look at it and see if I can get it working.

BTW, congrats on the promotion!

Cheers,
Steve

**Comment by ballsteve** [comment]
I've just checked-in a very crude entity expansion feature to the dev branch.

It could be a lot better... it basically works by making a new copy of the whole document. It would be better to just mutate the parts of the tree that are references. But you have to start somewhere!

**Reply by Devasta** [reply]
Hi Steve,

I'm back for a bit before another month end claims me. I'll keep looking at this. :)

**Comment by ballsteve** [comment]
I just read https://github.com/raphlinus/pulldown-cmark/blob/eb60cb976a12fb99972ddfc9b60cc1c6b20e096c/README.md#why-a-pull-parser
and it got me to thinking that perhaps a better strategy for entity expansion would be to convert the XML parser to a pull parser.

That is, the parser is an iterator. At each step in the iteration, if the production is an entity definition then it can be stored in a HashMap; if the production is an entity reference it can be expanded.

It might also be possible to set it up so that the parser pulls the input data. That way the body of the entity to be expanded could be injected into the inout stream.

This is going to be important at some point in the future to build the streaming features in XSLT.

**Reply by Devasta** [reply]
Nom does have streaming parsers that provide an incomplete status, but since it uses &str internally I don't think it does much except return incomplete instead of error. I've been working on getting rid of Nom entirely, rewriting the combinators to something we could chop and change to suit our needs, at the very least something that I can use to rturn a line and character number for errors in XML documents.

I'll be honest, I have never used streaming in XSLT 3, the files I work with have never been big enough, so I'll need to do some more reading.

**Reply by ballsteve** [reply]
No, I've never used XSLT 3 streaming either but if the entire spec is to be implemented then at some point it will have to be dealt with.

The largest single XML document I've ever had to process was almost 1GB in size. That was for a data migration project. It was difficult, but I solved the problem by buying more RAM for my laptop.

In my last job I regularly processed PBS XML documents (https://www.pbs.gov.au/downloads/2022/07/2022-07-01-xml-V3.zip) that are 300-350MB.

I now have a laptop with 32GB RAM so it easily ingests documents of these sizes by parsing the whole thing as a tree ;-)

**Reply by Devasta** [reply]
My job involves sending trading data to hedge fund prime brokers for settlement, so when the source document gets too big I usually adjust parameters so that smaller files generate more frequently. 😅

I'll look into streaming parsers and see what's available, and I'll continue working on what I am looking into in the meantime.

Licensing concerns (this using Apache Vs Saxon using GPL) means that I don't think we can look directly at the code for that for insight, but if you know C++ might be worth looking at Xerces?

**Reply by Devasta** [reply]
Created a new branch so you could see what I was playing around with, its passing all the tests the old one did but have not dived into full expansion support yet. I wouldn't merge it any time soon though, I want to make sure its rock solid before that happens, if it happens at all.

**Reply by Devasta** [reply]
Right, hoping to take a stab at getting this more complete this weekend.

Any idea what would be the best approach for the args for these combinators? At the moment I have

   type ParseResult<Output> = Result<(String, usize, Output), usize>;

Which works well enough, but if I change this to include DTDs and a configuration context

   type ParseResult<Output> = Result<(String, usize, CFG, DTD, Output), usize>;

Then it may become a bit unweildy, especially if we want to use them for something like XQuery way way down the line.

**Reply by ballsteve** [reply]
I just looked over the code in the branch... impressive! It looks like the parser functions return a function which is then used to parse the input.

As far as the ParseResult arg goes, adding more arguments into the result for different components does seem to be unwieldy. What happens when, inevitably, there are more components to be parsed?

**Reply by Devasta** [reply]
A few updates made now. There is a struct called "ParserConfig" which will eventually have some user supplied args, but for now is carrying some configurations from one parser to the next as well as the DTD.

Defense against the billion laugh attack is basic, we track the index where the expanded entity ends, as well as current entity depth. We reset depth counter whenever we go past that index.

I've got nothing for document validation, tbh I don't even know where to begin with that, and until I do I'm just pulling the entity declarations into strings, nothing fancy.

Still need to work out an answer for streaming, I am hoping that the combinators I have made here could still be used on them, but more reading required.

**Reply by ballsteve** [reply]
An update of my own:

I've found a new way to construct and navigate trees. This involves two phases: firstly, construct the tree into a mutable, but not fully navigable, structure. Then, convert it into an immutable, but fully navigable, tree. The latter uses Weak references to represent child->parent and sibling relationships.

Since this process involves two phases, I was thinking of integrating it into the parsing/expansion subsystem. I'm not sure this will work entirely seamlessly, but I'm going to try some experiments to see how far we get.

**Reply by Devasta** [reply]
Go for it! And if you find that the stuff I'm playing around with is causing more problems than it fixes, let me know and we can look at other solutions.

There are projects like quick-xml out there that appear to parse limited subsets of XML pretty quick, and stream it too, so could take cues from those if we need to.

**Reply by Devasta** [reply]
Hi Steve,

So, I've been playing around with a few different things for the XML parser, what I currently have is unproven but I think it could work: Basically, instead of having a &str that can be carried through the combinators, it has a struct that is carrying a .chars() iterator and a vec<char> that will contain expanded entities. I have implemented .next() on this struct to pop a char off the vec, unless its empty in which case it grabs the next char from the iterator.

Currently at "1801 passed; 709 failed; 127 ignored; 0 measured; 1 filtered out; finished in 1.07s". Seems to work very fast on the unit tests, but things get slow when the file sizes increase. An XML document that contains a million elements is taking almost a minute to process in release mode, but I am hoping I can find the bottleneck... assuming there is one and its not just the whole design is bad.

I've left it way too long since I last merged stuff in, so I am going to merge my ParserDev branch with your work on everything else. Can you take a look and let me know what you think of what I've got?

**Reply by ballsteve** [reply]
I will take a look and get back to you.

In the meantime, I've been implementing the intmuttree (Interior Mutability Tree) module across the package. Just got the XSLT module using Item::Node trait, but now need to resolve logic errors.

One thing that has popped up is that the XML parser only needs a single pass to create the document, which is a good thing, but it can't resolve XML Namespace declarations and usage during parsing. I've ended up having to make another recursive descent of the parsed tree in order to fix namespace usage. Ugh! Perhaps your new parser can also build state and pass that forward through the parsing process to be able to match namespace prefixes to namespace URIs as it goes?

**Reply by Devasta** [reply]
I'll give it a try. Looks like I should have tried to merge earlier, I can see these RNodes are in use now, will modify to use those if what I've got looks promising.

**Comment by Devasta** [comment]
I've got my parser down to 26 seconds for that large file, still running at like 20% the speed of the original but I'm happy all the same.

I can try and resolve namespace declarations as I go along, if we think the parser I have made has any promise. The iterator already carries the entities along as it goes down the document, could try to have it carry the namespaces.

**Reply by Devasta** [reply]
As an aside, how do you run the benchmarks stuff you have in the project?

**Comment by ballsteve** [comment]
Hi Daniel,

That sounds like great progress!

Dealing with namespaces would be a big win. Currently resolving namespaces requires a second pass through the document.

Cheers,
Steve

**Comment by ballsteve** [comment]
Hi Daniel,

I’ve adopted Criterion for benchmarking. It seems to work really well.

Cheers,
Steve

**Comment by Devasta** [comment]
Right, so looking at the Dev branch and the ParserDev branch, the two have diverged quite a bit (my fault for letting it go months without merging, future changes will be far more incremental) and it won't be a simple merge back together, I'm going to try and sort that first.

**Comment by ballsteve** [comment]
I've approved ParserDev3 and merged it into the dev branch. You can now delete the ParserDev3 branch.

The good news is that, indeed, the XSLT tests all pass. The bad news is that there are a lot of XML compliance tests now failing :-(

**Reply by Devasta** [reply]
Working through them now. A lot of tests that are failing come down to DTD support. I need to start looking at validation and figure out a solution for external DTDs.

One thing I was considering, for popular DTDs like XHTML and Docbook and so on, we should probably bundle a local copy in the project?

**Comment by ballsteve** [comment]
Do we want to support DTD validation? I’m happy to skip it. If someone wants validation then they can implement XSD or Relax-NG.

Regarding including DTDs, like XHTML or DocBook, that sounds like a “batteries included” approach. It’s a slippery slope ;-)

Cheers,
Steve

**Comment by Devasta** [comment]
Skipping DTD validation would mean skipping a lot of the tests, which is fine by me. 😄

Since XHTML has entities like &amp;copy; and so on, we can either bundle them in the project or look into external DTD support. Either is grand I think, but anything that allows us to skip network calls will help with kicking cans down the road when it comes to security. 😅

**Comment by Malavos** [comment]
GitHub said I was mentioned in this discussion, but I don't see anything mentioning me. Is GitHub drunk, or there was a spammer who comment was deleted?

**Reply by Devasta** [reply]
Hey!

There was a spammer whose comment was deleted, but you are welcome to join the discussion all the same. 😄
